### PR TITLE
RMB-938: Provide postgres-testing containers with read/write split

### PR DESCRIPTION
**Pull request by steveellis**
Implements streaming replication from primary to new read-only standby container. Changes `PostgresTester` interface and `PostgresTesterContainer` implementation to include a new host and port for the read-only standby to allow testing of read/write split. Writes new unit test to test replication and read only behavior on standby.

### CODE REVIEW

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 1]
```diff
@@ -4348,6 +4349,10 @@ public void startPostgresTester() {
     }
     postgreSQLClientConfig.put(PORT, postgresTester.getPort());
     postgreSQLClientConfig.put(HOST, postgresTester.getHost());
+
+    // TODO If we enable these 8 tests fail.
+    postgreSQLClientConfig.put(HOST_READER, postgresTester.getReadHost());
+    postgreSQLClientConfig.put(PORT_READER, postgresTester.getReadPort());
```
Comment (steveellis) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:4355]: @julianladisch Two tests fail (not 8 as the comment states). These failed tests are `PostgresClientIT.getNewInstanceSharedPool` and `GeneralFunctionsIT.upsert`. Any insight as to what could be happening here?

Comment (steveellis) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:4357]: @julianladisch I fixed both of the tests that were failing. For the upsert test fix, if I get the connection from `getSQLReadConnection` rather than relying `withReadConnection` it succeeds. The only difference I can see is that one news up a new `Conn` object.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 2]
```diff
@@ -34,44 +44,143 @@ public PostgresTesterContainer() {
...
+    var replicationUser = "replicator";
+    var primaryHost = "primaryhost";
+    var dataDirectory = "/var/lib/postgresql/standby/";
+    var tempDirectory = "/tmp/standby/";
+    var hostVolume = "/tmp/rmb-standby-" + UUID.randomUUID();
+    var network = Network.newNetwork();
+
+    // In postgres replication the default is for commit to be async. So no configuration is needed, but our
+    // default for tests is to configure it for synchronous commit.
+    boolean configureSynchronousCommit = System.getProperty(POSTGRES_ASYNC_COMMIT) == null;
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:62]: I hear what you're saying, but think handling this with a system property is a good approach for now. Yes it could be handled differently, by making it part of the constructor or properties on the class, but then we would need to rewrite a lot of test code, which will take time that we don't have right now, without providing a lot of immediate value.

What I do think we need right now is a simple way to run all of the tests in either sync or async mode, and this gives us that.

Here's how I'm imagining this system property could be used. Say a team is having trouble with the read/write split in their code and they have traced this to one of RMB's methods. Say the problem is because replication is async in their prod/test env (which is very likely going to be the case). They can now verify and replicate this problem locally by setting this flag, running their test, and potentially submit a PR that they know will fix it because it can be replicated and tested locally.

The point is that right now (with this PR) we don't need to fix every problem that comes with async replication, but this property gives teams a simple way to solve them.

Another way this property could be used (eventually) is with automation. Jenkins could do two passes of `mvn test` to test all of the tests with either configuration or perhaps other configurations that emerge that we'd like to test. I say _eventually_ because, right now not every test passes with async. In my testing two tests intermittently fail for example. Is that a problem? I say no not today. Let's get this out there so teams can use it and decide how important it is to fix every test later. We can always open new tickets.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 3]
```diff
@@ -34,44 +44,143 @@ public PostgresTesterContainer() {
...
+      var waitForHbaRelooad = Wait.forLogMessage(".*database system is ready to accept connections.*", 1);
+      primary.waitingFor(waitForHbaRelooad);
+
+      // Create replication user.
+      var createReplicationUser = "CREATE USER " + replicationUser + " WITH REPLICATION PASSWORD 'abc123'";
+      primary.execInContainer("psql", "-U", username, "-d", database, "-c", createReplicationUser);
+
+      // Make a temporary container that only gets used to generate the data directory, which we bind to the host
+      // file system. Don't use PostgreSQLContainer for this because it will start a db and make it very hard to clear
+      // out the data dir which is required for pg_basebackup.
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:95]: The problems that the temp container solves don't have to do with running pg_basebackup per se, even though were running it there. The temp container solves two problems that are really related to using TestContainers for what we're doing. These "problems" are:
* We need run pg_basebackup in  `GenericContainer` rather than the `PostgresSQLContainer` because the latter starts pg, and effectively locks the data directory because it starts pg. We need a container that has PG but doesn't start it, hence the generic one.
* Because we're using a `GenericContainer` and because pg isn't started, test containers won't expose any ports on the container, and even if we did it wouldn't start, because TestContainers really wants a process running at the port, otherwise it again won't start.

There could be a way to do it with two containers, but I spent a lot of time trying to make it work and the only workaround I could find was what we have with the filesysystem mount. Let's keep it.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 4]
```diff
@@ -34,44 +44,143 @@ public PostgresTesterContainer() {
...
+      primary = new PostgreSQLContainer<>(dockerImageName)
+          .withDatabaseName(database)
+          .withUsername(username)
+          .withPassword(password)
+          .withNetwork(network)
+          .withNetworkAliases(primaryHost)
+          .waitingFor(Wait.forLogMessage(".*database system is ready to accept connections.*\\n", READY_MESSAGE_TIMES));
+
+      if (configureSynchronousCommit) {
+        primary.withEnv("PGOPTIONS", "-c synchronous_commit=remote_apply");
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:74]: In line 125. The standby name is the default which is `walreciever` but to enable sync you still have to set it and reload.If I set this in PG_OPTIONS the container doesn't start probably because it expects the standby to be available which it isn't yet.

File: domain-models-runtime/src/test/java/org/folio/rest/persist/GeneralFunctionsIT.java [hunk 5]
```diff
@@ -21,22 +21,24 @@ public class GeneralFunctionsIT extends PostgresClientITBase {
...
-      }));
-    }));
+        .onComplete(context.asyncAssertSuccess(upsert -> {
+          assertThat(sql, upsert.getUUID(0), is(id));
+          postgresClient.getSQLReadConnection(conn -> {
+            postgresClient.getById(conn, "t", id.toString(), context.asyncAssertSuccess(get -> {
+              assertThat(get, is(json));
+            }));
+          });
+        }));
```
[the remark this answers is not in the document]
Comment (steveellis) [reply domain-models-runtime/src/test/java/org/folio/rest/persist/GeneralFunctionsIT.java:33]: The strangeness might be due to [what I observed in my ](https://github.com/folio-org/raml-module-builder/pull/1123/files#diff-7dce7229cecb42a85babe8bc265ecd46f1f32cfa02089ff29a7e3cdfcba6ec43R112)`testReadWrite` test. Even with sync configured, the first write has to have that `synchronous_commit` flag set in a transaction in that test. We're starting up `PostgresTesterContainer` quite a few times where this could catch us. But once the first data replicates, then subsequent operations are fully synchronous. Probably because the  WAL has to be "woken up" or something in the first transaction. In any case the solution is probably to force the test to use the read/write conn either way.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:97]: > Can you explain in more detail why the straightforward way doesn't work?
> 
> [master...RMB-938-container-read-write-split-julian](https://github.com/folio-org/raml-module-builder/compare/master...RMB-938-container-read-write-split-julian)

What you have does appear to work but you already know that I was trying to do everything by `execInContainer` and ran into trouble with perms which you have solved with a `Transferable` and the `fileMode` param, which I did not know about. But what is the magic of the `docker-entrypoint-initdb.d` directory?

File: util/src/main/java/org/folio/util/PostgresTester.java [hunk 6]
```diff
@@ -9,7 +10,7 @@ public interface PostgresTester extends Closeable {
    * @param username Postgres username
    * @param password Postgres password
    */
-  void start(String database, String username, String password);
+  void start(String database, String username, String password) throws PostgresTesterStartException;
```
[the remark this answers is not in the document]
Comment (steveellis) [reply util/src/main/java/org/folio/util/PostgresTester.java:13]: I added `PostgresTesterStartException` because of a sonar to consolidate `IOException` and `InterruptedException` which were needed because of my use of `execInContainer` but your branch doesn't use `execInContainer` so there's no need for any `throws` items on the method.

File: postgres-testing/src/test/java/org/folio/postgres/testing/PostgresTesterContainerTest.java [hunk 7]
```diff
@@ -48,4 +58,70 @@ public void testGetHost() {
...
+      Assert.assertEquals("ERROR: cannot execute CREATE TABLE in a read-only transaction", exception.getMessage());
+
+      // Should we test synchronous commit? The default is true.
+      boolean testSynchronousCommit = System.getProperty(PostgresTesterContainer.POSTGRES_ASYNC_COMMIT) == null;
+      if (!testSynchronousCommit) {
+        return;
+      }
+
+      // The first writes to the replicated cluster requires remote_apply. Subsequent ones don't.
+      stmt.executeUpdate("BEGIN; SET LOCAL synchronous_commit = remote_apply; CREATE TABLE crew (id SERIAL PRIMARY KEY, name VARCHAR(50)); COMMIT;");
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/test/java/org/folio/postgres/testing/PostgresTesterContainerTest.java:113]: Well in your branch you have removed this and it appears to work without it, probably because you are setting the synchronous_standby_names on init of the primary. Fine and good.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:97]: I see that docker-entrypoint-initdb.d is the standard place for db containers to put sql and such for Postgres and mysql. New to me.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 8]
```diff
@@ -1,16 +1,30 @@
...
+  public static final String PRIMARY_ALIAS = "postgresprimary";
+  public static final String STANDBY_ALIAS = "postgresstandby";
+  public static final String POSTGRES_ASYNC_COMMIT = "PG_ASYNC_COMMIT";
 
-  static public final String DEFAULT_IMAGE_NAME = "postgres:12-alpine";
+  private static final int READY_MESSAGE_TIMES = 2;
 
-  private PostgreSQLContainer<?> postgreSQLContainer;
+  private static final Logger LOG = LoggerFactory.getLogger(PostgresTesterContainer.class);
+  private static final int FILE_MODE = 493; // 493 is the decimal of octal 0755.
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:21]: I know right? Sonar [doesn't like octals.](https://rules.sonarsource.com/java/type/Code%20Smell?search=Octal) (S1314). Would you prefer to suppress the warning, ignore it, or I could do something like `Integer.valueOf("755", 8)`which would probably be better than commenting it as I have done.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 9]
```diff
@@ -35,43 +65,113 @@ public PostgresTesterContainer() {
...
+        "echo 'synchronous_commit=remote_apply' >> /var/lib/postgresql/data/postgresql.conf\n"
+            + "echo \"synchronous_standby_names='*'\" >> /var/lib/postgresql/data/postgresql.conf\n"
+            + "echo 'host replication replicator 0.0.0.0/0 trust' >> /var/lib/postgresql/data/pg_hba.conf\n"
+            + "psql -U \"$POSTGRES_USER\" -d \"$POSTGRES_DB\" -c 'CREATE USER replicator WITH REPLICATION'");
+
+    if (configureAsyncCommit) {
+      replicationSh = Transferable.of(
+          "echo 'host replication replicator 0.0.0.0/0 trust' >> /var/lib/postgresql/data/pg_hba.conf\n"
+              + "psql -U \"$POSTGRES_USER\" -d \"$POSTGRES_DB\" -c 'CREATE USER replicator WITH REPLICATION'");
+    }
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:84]: I can't agree. I have a hard time seeing the value of this PR if it only provides synchronous replication. Why not just keep a single container then? Just throwing an error if you try to write to the reader is somewhat useful, but being able to test  async is more useful IMO.

My opinion is that _any_ FOLIO code that assumes that a write is immediately available in a read replica should be revisited. Having this option gives us the ability to test these necessary changes. The solution is simple, if you need the data right away, use the read/write endpoint not the reader one.

Asynchronous replication is the default for postgres replication and requires extra configuration to make it synchronous. Also if hosters use AWS RDS, which many do, read replica replication is _always_ asynchronous. This results in unexpected replication behavior in production.

From: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_ReadRepl.html

image

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 10]
```diff
@@ -1,16 +1,30 @@
...
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.images.builder.Transferable;
+import org.folio.util.PostgresTester;
 
 public class PostgresTesterContainer implements PostgresTester {
+  public static final String DEFAULT_IMAGE_NAME = "postgres:12-alpine";
+  public static final String PRIMARY_ALIAS = "postgresprimary";
+  public static final String STANDBY_ALIAS = "postgresstandby";
+  public static final String POSTGRES_ASYNC_COMMIT = "PG_ASYNC_COMMIT";
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:29]: I'm not loving the idea of the interface's implementation being documented in the interface itself (if I'm understanding). That doesn't seem right. What I did do is put a javadoc on this property here in the implementation. See what you think.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java [hunk 11]
```diff
@@ -35,43 +65,113 @@ public PostgresTesterContainer() {
...
+    if (standby == null) {
+      throw new IllegalStateException("read only not started");
+    }
+    return standby.getHost();
+  }
+
+  /**
+   * On the Docker network the hostname aliases are {@link #PRIMARY_ALIAS} and {@link #STANDBY_ALIAS},
+   * they listen on port 5432.
+   */
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java]: The interface already has a javadoc for this method. I can delete the javadoc here though. I don't think the `@link` items belong in the interface though.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:107]: But do we have the power to say that everyone absolutely has to run sync replication in their config? If we did I would agree with you, but as noted above I don't think we do have this power and should not try to impose it.

File: domain-models-runtime/src/test/java/org/folio/rest/persist/GeneralFunctionsIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply domain-models-runtime/src/test/java/org/folio/rest/persist/GeneralFunctionsIT.java:33]: The test doesn't pass without my change in sync or async mode. I don't fully understand why. The only difference between the code I have, and the original code that I can see is that the one that works news up the `Conn` object if I'm remembering right.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:107]: I won't disagree with the points you're making but I don't see why it should stop us from having an *option* to test an async configuration in RMB. Lets say we're only going to support eventual consistency in new APIs, as you suggest. Wouldn't it be nice to have a way to test those APIs? That's all that is at stake here IMO. I think whether we have to revisit old code is out of scope for this.

File: postgres-testing/src/test/java/org/folio/postgres/testing/PostgresTesterContainerTest.java [hunk 12]
```diff
@@ -48,4 +60,99 @@ public void testGetHost() {
...
+          });
+    }
+    tester.close();
+  }
+
+  private void verify(Callable<Boolean> callable, boolean isAsync) throws Exception {
+    if (!isAsync) {
+      assertTrue(callable.call());
+    }
+    await().atMost(5, TimeUnit.SECONDS).until(callable);
```
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/test/java/org/folio/postgres/testing/PostgresTesterContainerTest.java:155]: I thought I had a return statement after assertTrue. But your change is fine as well.

File: postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:108]: > The default for replicas must be synchronous because all existing code relies on consistency.

The default _is_ synchronous here, not async. Not tests will fail and people looking at it can be referred to the javadoc. I don't see a risk to providing the async option as an experimental feature.
 
> Therefore async mode can only been used on a per transaction basis.

Async mode _could_ only be done on a transaction basis, except that for anyone using RDS with read replicas and who wants to take advantage of the split, it is configured on the db level, not on the transaction level, and is something they can't change.

A lot of time has been invested in this read write split, and implementers who are using RDS want to take advantage of read replicas, and they should at least have the option of exploring what would be needed to make that work. Without this they don't have that option. Why shut down the _possibility_ of RDS users eventually of taking advantage of the split?

I agree given your assessment of the current code that deploying read replicas in RDS should be done with caution. But the read write split is already merged. We should now provide an option for better testing of this configuration.

[the remark this answers is not in the document]
Comment (mtraneis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:108]: > Most module code that uses PostgresClient has been written before RMB added the reader option.
> 
> It's in the contract of the PostgresClient methods that written values are immediately available, without eventual consistency

@julianladisch, do the R/W split changes that were implemented on the existing RMB readonly methods continue to uphold this warranty even if the underlying database technology like RDS does not replicate synchronously? 

> There are use cases where performance is important, these can be implemented at new (!) APIs that clearly document that there is only eventual consistency. Code that uses these new APIs is aware of eventual consistency and can take measures to avoid working with inconsistent data.

Here I don't think it matters if the client uses the new APIs or the old APIs because RMB does not have control over how the DB replicates the data. I agree with Steve that there is a need for clients to test their code with the async replication option when R/W split mode is enabled so that they would know all the effects that enabling R/W split could bring.  If unit testing with RMB does not inform clients that there might be an issue due to replication, when they deployed FOLIO into the real world where replication is asynchronous, not until then will an issue be found and it will be too late.  Moreover, if a team has to troubleshoot this kind of an issue (due to async replication) and they cannot reproduce the problem with unit tests, are we not making it more difficult for people to consume this code?  

I agree with you that folks should know what the capabilities of RMB are, what is guaranteed or not, and these things should be documented clearly in the README and elsewhere as appropriate.  It seems to me that not having this option for teams to test (with async replication) will do them a disservice.

[the remark this answers is not in the document]
Comment (mtraneis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:108]: > Having an async replica in the postgres-testing containers is not sufficient to show that the code under test actually supports eventual consistency resulting from async replication.

Thanks, @julianladisch . I was not suggesting that RMB or our client code must prove that they support eventual consistency. I think Steve concurs with me that we'd want to expose a potential problem due to async replication if one exists. 

> In addition updates must be hold back so that the replica has stale data, this enforced stale data is lacking from the current implementation of PG_ASYNC_COMMIT option. The time lag of the async replica can increase under load in production, therefore it is not sufficient to test with the small default time lag.

So basically you're saying that having this configureAsyncCommit configuration option is useless even to simulate async replication?

[the remark this answers is not in the document]
Comment (steveellis) [reply postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java:118]: > Having an async replica in the postgres-testing containers is not sufficient to show that the code under test actually supports eventual consistency resulting from async replication. In addition updates must be hold back so that the replica has stale data, this enforced stale data is lacking from the current implementation of PG_ASYNC_COMMIT option. The time lag of the async replica can increase under load in production, therefore it is not sufficient to test with the small default time lag.

I think this is a valid point Julian, so I looked into how we can introduce a simulated replication delay in the config. There is `recovery_min_apply_delay` which does what we want. I have added it to the async config, introduced a new constant called `SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS` and modified the async version of the unit test to make use of it like this:
```
await()
    .atLeast(PostgresTesterContainer.SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS - 100, TimeUnit.MILLISECONDS)
    .atMost(PostgresTesterContainer.SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS + 100, TimeUnit.MILLISECONDS)
    .until(...);
```
Does this address your concern?
